### DISCUSSION TITLE
:class not evaluating as expected

### DISCUSSION THREAD

**Discussion by jkrk678** [alpinejs/alpine#3386] [7. Bugs (archived)] [answered]
I am trying to write a TableSorter component using some principles of state machines. All of my code seems to be working as I expect, but I have a :class directive that is not doing what I expect. To check my expression, I am also logging the value of the expression, which does give me what I expect. Any guidance on what I'm doing wrong? 

https://codepen.io/jkrk678/pen/zYLpNwY

**Comment by ekwoka** [comment, marked as answer]
I can identify some of the issues, but still having some problems, your system is quite different from how I would do it so I'm trying to solve without just rewriting in my style.

For starters, the reason it's not working is that you are referencing the object directly with `tableSort.blah` in the methods of the object. This won't work since that will be referring to the underlying object, while alpine will be listening for changes that pass through the proxy layer (which will be `this` in the functions). this is the reason why the functions themselves can get the correct result (they access and pass through the proxy to the object at the bottom) but the reactive listeners aren't triggered when setting new properties. (you can use an effect to log out the state whenever it changes and see it never reruns.)

After making this change though I'm getting that `this` is undefined which is really weird.

**Reply by ekwoka** [reply]
Looks like `this` in nested methods isn't working how I would have expected...may be bug?

for now, this works:
```js
import Alpine from 'https://esm.sh/alpinejs@3.10.5'
Alpine.data('TableSorter', () => ({
        data: {col: '', target: ''},
        state: 'default',
        sort(e) {
                console.log(this)
                // set state
                this.data.target = e.target.dataset.col;
                const stateTransitionHash = {
                    default: 'ascending',
                    ascending: this.targetIsSortedColumn ? 'descending' : 'ascending',
                    descending: this.targetIsSortedColumn ? 'default' : 'ascending'
                };
                this.state = stateTransitionHash[this.state];

                // assign column
                this.data.col = this.data.target;
            },
            get targetIsSortedColumn() {return this.data.target === this.data.col;},
            get targetIsNotSortedColumn() {return this.data.target !== this.data.col;},
    })
);

Alpine.start();
```

have all methods/getters on the top level, nested ones were refering only to their parent object. I feel like this wasn't the case before...

**Reply by ekwoka** [reply]
Actually, idk, this worked fine:

```js
test: {
      foo() { console.log(this.test.foo) }
    }
```
    
 but for some reason you `method` object and `guards` object both didn't work.

**Reply by ekwoka** [reply]
As far as style, I'd probably change the `sort` to accept specific data instead of using the `dataset` to do it. Additionally, you might want to use `currentTarget` instead of `target` to ensure you get the correct element (for clicks, `target` is the literal lowest element clicked on, but `currentTarget` is the element the event is actually listening on).
